clickhouseexporter: add traces support

hanjm opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
Support export OTLP traces to ClickHouse.
**Link to tracking Issue:** <Issue number if applicable>
#8028 #10300 
**Testing:** <Describe what testing was performed and which tests were added.>
1. unit test 
2. test with example

hanjm committed:
clickhouseexporter: add traces support

hanjm committed:
add changelog yaml

hanjm committed:
fix goimport

hanjm commented:
hi @dmitryax , Could you help review, Thank you.

hanjm commented:
hi @bputt-e @alvinhom would you like to help review? Thanks.

bputt-e commented on exporter/clickhouseexporter/README.md:
```diff
@@ -1,117 +1,256 @@
 | Status                   |           |
 | ------------------------ |-----------|
 | Stability                | [alpha]    |
 | Supported pipeline types | logs      |
```
should this include spans

hanjm replied:
fixed, thanks

bputt-e commented on exporter/clickhouseexporter/README.md:
```diff
@@ -1,117 +1,256 @@
-> ClickHouse is an open-source, high performance columnar OLAP database management system for real-time analytics using SQL.
-> Throughput can be measured in rows per second or megabytes per second. 
-> If the data is placed in the page cache, a query that is not too complex is processed on modern hardware at a speed of approximately 2-10 GB/s of uncompressed data on a single server.
+This exporter supports sending OpenTelemetry logs to [ClickHouse](https://clickhouse.com/). It will also support spans
```
should this be updated to support logs and spans

alvinhom reviewed (commented):
How are the traces going to be consumed?  I see the traces will be stored in Clickhouse and you can probably use Grafana Clickhouse datasource to retrieve it, but it can only be used like a flatten table.  I would think we need to implement either a Tempo or Jaeger http endpoint so the traces can be visualized using either Grafana (Tempo/Jaeger datasource) or in Jaeger.

Why is there a need for a s_trace_id_ts table?  Will it help in retreiving the traces?

bputt-e commented:
@alvinhom 

We planned on either:
1. running tempo alongside clickhouse (clickhouse for the discovery phase, a replacement for loki)
2. Creating a tempo-compatible api that reads from clickhouse if the option above became too expensive

Considering Jaeger's clickhouse integration table schema https://github.com/jaegertracing/jaeger-clickhouse#how-it-works , maybe this exporter could benefit from a output format type (jaeger vs other) so that people have the option to choose the eventual backend.

alvinhom commented:
I would think option 2 is probably the way to go.  The benefit of using Clickhouse over Tempo/Elastic backend is definitely the discoverability and analytics it can offer.  But I think the sizing aspect (both space and ingestion time) is also an important reason for choosing Clickhouse.

hanjm commented:
> How are the traces going to be consumed? I see the traces will be stored in Clickhouse and you can probably use Grafana Clickhouse datasource to retrieve it, but it can only be used like a flatten table. I would think we need to implement either a Tempo or Jaeger http endpoint so the traces can be visualized using either Grafana (Tempo/Jaeger datasource) or in Jaeger.

It is easy to implement a basic http api to find trace by traceID, for trace UI provider like grafana https://grafana.com/docs/grafana/latest/explore/trace-integration/, it is out of opentelemetry-collector scope.

> Why is there a need for a s_trace_id_ts table? Will it help in retreiving the traces?

There are two mainly scenario about query spans :
1. query by serviceName or spanName, with attribute condition or other optionally condtition.
2. query by traceID.

the main table `otel_traces` is mainly order by `serviceName` 、`spanName` 、 `timestamp`, the primary key fields is order by cardinality count for merge tree mechanism, so it will be benifit for the query by  `serviceName` or spanName and time range.
but for the query traceID scenario, the `traceID` is high cardinality and randomly value, so it is will not benifit query only traceID scenario. 

the `otel_traces_trace_id_ts` table  is the key design, it is a small table order by `traceID`, write by materialized view, so it is very fast to query the start time and end time for a traceID, then query the `otel_traces` table by traceID with time range effectively.
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/7bc5de66b07ad0eb2b048484c4b11e9466603e2a/exporter/clickhouseexporter/README.md?plain=1#L353-L372

usually the span time range is short, it will speed up significantly when spans volume is large.
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/7bc5de66b07ad0eb2b048484c4b11e9466603e2a/exporter/clickhouseexporter/README.md?plain=1#L144-L168

another way to speed up query by traceID is create materialized view store the full spans order by traceID , but it is more expensive for write performance and space consume, and not improve more significantly.

bputt-e commented:
@hanjm I added my review comments, only one comment referred to actual logic in your code, the rest were documentation related

hanjm committed:
update readme.md

hanjm commented:
> @hanjm I added my review comments, only one comment referred to actual logic in your code, the rest were documentation related
@bputt-e Thank you, fixed.

hanjm committed:
fix rebase main

bputt-e commented:
@hanjm for batching, assuming we're using the batch processor, isn't this still doing single inserts into clickhouse? Would adding in batching via the go client be more performant or is the client sending inserts in "batches" to prevent network overhead

hanjm commented:
> @hanjm for batching, assuming we're using the batch processor, isn't this still doing single inserts into clickhouse? Would adding in batching via the go client be more performant or is the client sending inserts in "batches" to prevent network overhead

@bputt-e not need. the batch processor do the batch to one ptrace.Traces, then the exporter will do one batch insert, for clickhouse sdk driver, it will do batch insert behaind the PrepareContext.  https://github.com/ClickHouse/clickhouse-go/blob/main/examples/std/batch.go#L59
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ae3fcb5b6b94fa0582c719ac5b355a727be79438/exporter/clickhouseexporter/exporter_traces.go#L30-L33

dmitryax commented on exporter/clickhouseexporter/README.md:
```diff
@@ -1,117 +1,255 @@
-- `logs_table_name` (default = otel_logs): The table name for logs.
+- `ttl_days` (default= 0): The data time-to-live in days, 0 means no ttl.
+- `database` (default = otel): The database name.
+- `logs_table_name` (default = otel_traces): The table name for logs.
+- `traces_table_name` (default = otel_logs): The table name for traces.
```
Looks like default values are wrong here

hanjm replied:
Fixed. thanks.

dmitryax commented on exporter/clickhouseexporter/README.md:
```diff
@@ -1,117 +1,255 @@
-- `dsn` (no default): The ClickHouse server DSN (Data Source Name), for example `tcp://127.0.0.1:9000?username=user&password=qwerty&database=default`
-   For tcp protocol reference: [ClickHouse/clickhouse-go#dsn](https://github.com/ClickHouse/clickhouse-go#dsn).
-   For http protocol reference: [mailru/go-clickhouse/#dsn](https://github.com/mailru/go-clickhouse/#dsn).
+- `dsn` (no default): The ClickHouse server DSN (Data Source Name), for
```
Not for this PR, but why this option must be `dsn` and not `endpoint` as all other exporter configurations have?

hanjm replied:
Good advice, I will submit another PR use `endpoint` for consistent.
the `dsn` can be split to `endpoint` `database` `params`

dmitryax commented on exporter/clickhouseexporter/README.md:
```diff
@@ -169,5 +311,66 @@ CREATE TABLE otel_logs
+    `Duration`      Int64 CODEC (ZSTD(1)),
+    `StatusCode` LowCardinality(String) CODEC (ZSTD(1)),
+    `StatusMessage` String CODEC (ZSTD(1)),
+    `Events.Timestamp` Array(DateTime64(9)) CODEC (ZSTD(1)),
+    `Events.Name` Array(LowCardinality(String)) CODEC (ZSTD(1)),
+    `Events.Attributes` Array(Map(LowCardinality(String), String)) CODEC (ZSTD(1)),
+    `Links.TraceId` Array(String) CODEC (ZSTD(1)),
+    `Links.SpanId` Array(String) CODEC (ZSTD(1)),
+    `Links.TraceState` Array(String) CODEC (ZSTD(1)),
+    `Links.Attributes` Array(Map(LowCardinality(String), String)) CODEC (ZSTD(1)),
```
Does it represent 7 independent arrays? Can you please clarify why this data model is chosen? Why can't we use a nested data structure instead?

hanjm replied:
> Does it represent 7 independent arrays?

No, it just copy from the result of `show create table`. for clearly, I have update it to nested data structure

hanjm committed:
revert changes for log exporter

hanjm committed:
fix logs_table_name default value

hanjm commented on exporter/clickhouseexporter/exporter.go:
```diff
@@ -73,11 +74,11 @@ func (e *clickhouseExporter) pushLogsData(ctx context.Context, ld plog.Logs) err
 			logs := ld.ResourceLogs().At(i)
 			res := logs.Resource()
 			resAttr := attributesToMap(res.Attributes())
+			var serviceName string
```
OK. let me submit another PR later.

hanjm committed:
rebase main

hanjm committed:
fix example docker run issue

hanjm committed:
fix lint

hanjm committed:
remove mutatation original configuration when Validate

hanjm commented on exporter/clickhouseexporter/config.go:
```diff
@@ -37,8 +40,12 @@ type Config struct {
 	// TTLDays is The data time-to-live in days, 0 means no ttl.
 	TTLDays uint `mapstructure:"ttl_days"`
+	// Database is the database name parse from dsn.
+	Database string `mapstructure:"-"`
```
OK , i removed it. I will submit another PR use `endpoint` for consistent.
the `dsn` can be split to `endpoint` `database` `params`

hanjm commented on exporter/clickhouseexporter/config.go:
```diff
@@ -56,6 +63,11 @@ func (cfg *Config) Validate() (err error) {
+	if e != nil {
+		err = multierr.Append(err, fmt.Errorf("invalid dsn format:%w", err))
+	}
+	cfg.Database = strings.TrimPrefix(u.Path, "/")
```
fixed. thanks

dmitryax commented on exporter/clickhouseexporter/exporter.go:
```diff
@@ -177,24 +178,57 @@ var driverName = "clickhouse" // for testing
 	}
-	if _, err := db.Exec(query); err != nil {
-		return nil, fmt.Errorf("exec create table sql: %w", err)
+	if _, err := db.Exec(renderCreateTracesTableSQL(cfg)); err != nil {
+		return nil, fmt.Errorf("exec create traces table sql: %w", err)
+	}
+	if _, err := db.Exec(renderCreateTraceIDTsTableSQL(cfg)); err != nil {
+		return nil, fmt.Errorf("exec create traceIDTs table sql: %w", err)
+	}
+	if _, err := db.Exec(renderTraceIDTsMaterializedViewSQL(cfg)); err != nil {
+		return nil, fmt.Errorf("exec create traceIDTs view sql: %w", err)
```
Should we create both logs and traces tables if only one of the pipelines is configured?

hanjm replied:
Good point. we can add some config field about creating table for logs/traces (do_create/tablename/ ttl_days). let me create another pr about optimize config field)?

hanjm replied:
i think i can split into initLogsDB and initTracesDB in this PR.

hanjm replied:
Fixed in this PR, it will [create table on demond](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13442/commits/68217b3a0361b2c91aad03a75e1400d2e3df5751)

next PR i will add some config field about creating table.

hanjm committed:
create table on demond

dmitryax commented on exporter/clickhouseexporter/exporter.go:
```diff
@@ -28,30 +29,31 @@ import (
-func newExporter(logger *zap.Logger, cfg *Config) (*clickhouseExporter, error) {
+func newExporter(logger *zap.Logger, cfg *Config, createLogsTable, createTracesTable bool) (
+	*clickhouseExporter, error) {
 	if err := cfg.Validate(); err != nil {
 		return nil, err
 	}
```
This is not needed. Validate is called implicitly when receiver is created https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/moved_config.go#L77

hanjm replied:
OK , i remove it

dmitryax commented on exporter/clickhouseexporter/exporter.go:
```diff
@@ -28,30 +29,31 @@ import (
 	}
 
-	client, err := newClickhouseClient(cfg)
+	client, err := newClickhouseClient(cfg, createLogsTable, createTracesTable)
```
I believe it'd be cleaner if we have different factories `newLogsExporter` and `newTraceExporter` instead of one `newExporter`. Each of them creates databases after calling `newClickhouseClient`. `newClickhouseClient` should just create client, not databases

hanjm replied:
Thank you , I have refactored the exporter factories, please review.

hanjm replied:
hi @dmitryax, I have refactored the exporter factories, please review, thank you.

hanjm committed:
refactor exporter factories

hanjm committed:
rebase mian, use PutStr instead of PutString

hanjm committed:
fix changelog

dmitryax reviewed (approved):
Sorry for the delay
